File: android/src/main/java/com/paulingalls/realtimeaudio/BarGraphVisualizer.kt
package com.paulingalls.realtimeaudio

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import kotlin.math.abs
import kotlin.math.ceil


class BarGraphVisualizer : BaseVisualization() {
    private val paint = Paint()
    private var mainColor: Int = Color.BLUE

    override fun draw(canvas: Canvas, width: Float, height: Float) {
        val barCount = 20
        val minBarHeight = 20
        val barWidth = width / barCount
        val gap = 4.0
        val div = samples.size / barCount
        paint.strokeWidth = (barWidth - gap).toFloat()
        paint.color = mainColor

        for (i in 0 until barCount) {
            val bytePosition = ceil((i * div).toDouble()).toInt()
            val amplitude = abs(samples[bytePosition])
            val top = height - minBarHeight - (amplitude * (height - minBarHeight))
            val barX = (i * barWidth) + (barWidth / 2)
            canvas.drawLine(barX, height, barX, top, paint)
        }
    }

    override fun setColor(color: Int) {
        mainColor = color
    }
}

File: ios/BarGraphVisualizer.swift
import AVFoundation

class BarGraphVisualizer: BaseVisualizer {

  override func updateVisualization(with samples: [Float]) {
    let width = shapeLayer.bounds.width
    let height = shapeLayer.bounds.height
    let barCount: Int = 20
    let minBarHeight: CGFloat = height * 0.1
    let barWidth: CGFloat = width / CGFloat(barCount)
    let gap: CGFloat = 4.0
    let div: Int = samples.count / barCount

    let path = UIBezierPath()
    for i in 0..<barCount {
        let bytePosition: Int = Int(ceil(Double(i * div)))
        let amplitude: CGFloat = CGFloat(abs(samples[bytePosition]))
        let top: CGFloat = height - minBarHeight - (amplitude * (height - minBarHeight))
        let barX: CGFloat = (CGFloat(i) * barWidth) + (barWidth / 2)
        
        path.move(to: CGPoint(x: barX, y: height))
        path.addLine(to: CGPoint(x: barX, y: top))
    }
    shapeLayer.lineWidth = barWidth - gap
    shapeLayer.strokeColor = mainColor.cgColor
    shapeLayer.path = path.cgPath
  }
}
